src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs

```diff
@@ -360,59 +370,26 @@ private bool IsParsableFromString(ITypeSymbol type, out StringParsableTypeKind t
-            {
-                spec = GetOrCreateTypeSpec(type);
-
-                if (spec is null)
-                {
-                    RegisterUnsupportedType(type, descriptor);
-                    return false;
-                }
-
-                return true;
-            }
-
-            private EnumerableSpec? CreateArraySpec(IArrayTypeSymbol arrayType)
+            private EnumerableSpec? CreateArraySpec(IArrayTypeSymbol arrayTypeSymbol)
             {
-                if (!TryGetTypeSpec(arrayType.ElementType, Diagnostics.ElementTypeNotSupported, out TypeSpec elementSpec))
+                ITypeSymbol elementTypeSymbol = arrayTypeSymbol.ElementType;
+                if (!MemberTypeIsBindable(arrayTypeSymbol, elementTypeSymbol, Diagnostics.ElementTypeNotSupported, out TypeSpec elementTypeSpec) ||
+                    // We want a BindCore method for List<TElement> as a temp holder for the array values
+                    GetOrCreateTypeSpec(_typeSymbols.List.Construct(elementTypeSymbol)) is not EnumerableSpec listTypeSpec)
                 {
                     return null;
                 }
 
-                // We want a BindCore method for List<TElement> as a temp holder for the array values. We know the element type is supported.
-                EnumerableSpec listSpec = (GetOrCreateTypeSpec(_typeSymbols.List.Construct(arrayType.ElementType)) as EnumerableSpec)!;
-                RegisterTypeForBindCoreGen(listSpec);
-
-                EnumerableSpec spec = new EnumerableSpec(arrayType)
+                EnumerableSpec spec = new EnumerableSpec(arrayTypeSymbol)
                 {
-                    ElementType = elementSpec,
-                    ConcreteType = listSpec,
-                    InitializationStrategy = InitializationStrategy.Array,
+                    ElementType = elementTypeSpec,
+                    InstantiationStrategy = InstantiationStrategy.Array,
                     PopulationStrategy = CollectionPopulationStrategy.Cast_Then_Add, // Using the concrete list type as a temp holder.
-                    ToEnumerableMethodCall = null,
+                    TypeToInstantiate = listTypeSpec,
+                    PopulationCastType = null,
                 };
 
-                Debug.Assert(spec.CanInitialize);
-                RegisterTypeForBindCoreGen(spec);
-
-                return spec;
-            }
-
-            private bool IsSupportedArrayType(ITypeSymbol type)
-            {
-                if (type is not IArrayTypeSymbol arrayType)
-                {
-                    return false;
-                }
-
-                if (arrayType.Rank > 1)
-                {
-                    RegisterUnsupportedType(arrayType, Diagnostics.MultiDimArraysNotSupported);
-                    return false;
-                }
-
-                return true;
+                return TryRegisterTypeForBindCoreGen(listTypeSpec) && TryRegisterTypeForBindCoreGen(spec) ? spec : null;
```

Comment:
>TryRegisterTypeForBindCoreGen(listTypeSpec) [](http://example.com/codeflow?start=23&length=43)

should you try this check before creating `spec`?

Reply:
Given the new changes, I expect these methods to always return true. Replacing with debug asserts.